packages/components/src/tools-panel/tools-panel-item/component.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { useToolsPanelContext } from '../context';
+import { MENU_STATES } from '../utils';
+
+// This wraps controls to be conditionally displayed within a tools panel. It
+// prevents props being applied to HTML elements that would make them invalid.
+const ToolsPanelItem = ( {
+	children,
+	hasValue,
+	isShownByDefault,
+	label,
+	onDeselect,
+	onSelect,
+} ) => {
+	const {
+		checkMenuItem,
+		menuItems,
+		registerPanelItem,
+	} = useToolsPanelContext();
+
+	const isValueSet = hasValue();
+
+	useEffect( () => {
+		registerPanelItem( {
+			hasValue,
+			isShownByDefault,
+			label,
+			onDeselect,
+			onSelect,
+		} );
+	}, [] );
```

Comment:
Nice! 👍 

Have you checked what happens if one of those props is updated? Since this effect runs only once when the component is mounted, I guess those props wouldn't be updated in the context. But I'm not sure if it's a problem here.

Another thing to consider is when the component is unmounted. Will we keep the item registered in the context?

Reply:
I did run into an issue with those props not being updated in context. 

The problem was encountered when attempting to reset a block's attributes via `onDeselect`. Those attributes were effectively cached once `onDeselect` is registered through the highlighted initial hook. Even changing the `ToolsPanelItem` component to update the stored `onDeselect` or `onSelect` callbacks didn't solve the issue for me. 

The best [solution](https://github.com/WordPress/gutenberg/pull/32392/commits/f2851b64af8f73cc33c415b9ac005bec06eb5e20) I came across was for the callback itself to retrieve the latest block attributes and avoid the need to update the callbacks in the panel's context. 

Another option I explored was to make the `toggleItem` function in the `ToolsPanel` set a new `deselecting` state etc, which was passed through the context so that `ToolsPanelItem`s could execute their own callbacks. That required advising the `ToolsPanel` that the item had finished its work so that state could be cleared. That felt extremely hacky so it was abandoned before getting too deep into it.

Some better and fresh ideas would be greatly welcomed if you have any advice 🙏

Reply:
When testing the dimensions global styles using the `ToolsPanel` I ran into further issues. 

I've rolled back the [previous changes](https://github.com/WordPress/gutenberg/pull/32392/commits/f2851b64af8f73cc33c415b9ac005bec06eb5e20) to the dimensions block support hook and moved the execution of the `onDeselect` and `onSelect` callbacks down to the `ToolsPanelItem` itself. The previous menu item checked state is used in part to determine which callback should be run. 

This [new approach](https://github.com/WordPress/gutenberg/pull/32392/commits/3d7e73c4a3dd44f6fa65137dce4e9904bd80be9c) seems rather obvious after taking a moment to step back. The `ToolsPanel` use in the global styles and dimensions block support UI both work well for me now and the tests are still passing.